ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,534 @@
+        logger.info(
+            "Account to update with accountId={} accountName={} conflicts with an existing record with accountId={} accountName={}",
+            account.getId(), account.getName(), conflictingAccount.getId(), conflictingAccount.getName());
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * <p>
+   *   A cache to store {@link Account} information locally. It implements the {@link AccountService} interface,
+   *   and can respond to the queries for {@link Account}. Internally it maintains a {@link AccountTwoKeyMap} that
+   *   represents a group of {@link Account}s that can be indexed by both account id and name.
+   * </p>
+   * <p>
+   *   Operations on a {@link Cache} is thread-safe.
+   * </p>
+   */
+  private class Cache implements AccountService {
+    private final AtomicReference<AccountTwoKeyMap> accountTwoKeyMapRef = new AtomicReference<>();
+
+    /**
+     * Constructor that instantiates a cache with no data.
+     */
+    private Cache() {
+      accountTwoKeyMapRef.set(new AccountTwoKeyMap());
+    }
+
+    @Override
+    public Account getAccountById(short accountId) {
+      return accountTwoKeyMapRef.get().getAccountById(accountId);
+    }
+
+    @Override
+    public Account getAccountByName(String accountName) {
+      return accountTwoKeyMapRef.get().getAccountByName(accountName);
+    }
+
+    @Override
+    public Collection<Account> getAllAccounts() {
+      return accountTwoKeyMapRef.get().getAccounts();
+    }
+
+    @Override
+    public boolean updateAccounts(Collection<Account> accounts) {
+      throw new IllegalStateException("This method is not implemented, and is not expected to be called.");
+    }
+
+    @Override
+    public void close() throws IOException {
+      accountTwoKeyMapRef.set(null);
+    }
+
+    /**
+     * Atomically updates the cache with new id-to-account and name-to-account map. It assumes the {@link Account}s
+     * in the two maps have id and name one-to-one mapped.
+     * @param accountTwoKeyMap A {@link AccountTwoKeyMap} to replace the current one.
+     */
+    private void update(AccountTwoKeyMap accountTwoKeyMap) {
+      accountTwoKeyMapRef.set(accountTwoKeyMap);
+    }
+
+    /**
+     * Gets the {@link AccountTwoKeyMap} in this cache.
+     * @return The {@link AccountTwoKeyMap} in this cache.
+     */
+    private AccountTwoKeyMap getAccountTwoKeyMap() {
+      return accountTwoKeyMapRef.get();
+    }
+  }
+
+  /**
+   * <p>
+   *   A helper class as a representation of a collection of {@link Account}s, where the ids and names of the
+   *   {@link Account}s are one-to-one mapped. An AccountTwoKeyMap guarantees no duplicated account id or name,
+   *   nor conflict among the {@link Account}s within it.
+   * </p>
+   * <p>
+   *   Based on the properties, a {@code AccountTwoKeyMap} has two keys (index) to {@link Account}s: the accountId,
+   *   and the accountName.
+   * </p>
+   */
+  private class AccountTwoKeyMap {
```

Comment:
Facing multiple choices, finally renamed it as `AccountInfoMap`. Added javadoc to explain its property.